File: LeetCode/Study Plan/Practice Problems/Medium/Permutations/Java/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new LinkedList();

        ArrayList<Integer> numList = new ArrayList<>();
        for (int val : nums) numList.add(val);

        int size = nums.length;
        backtrack(numList, size, 0, res);
        return res;
    }

    private void backtrack(ArrayList<Integer> nums, int n, int ind, List<List<Integer>> res) {
        if (ind == n) {
            res.add(new ArrayList<>(nums));
            return;
        }
        for (int i=ind; i<n; i++) {
            Collections.swap(nums, ind, i);
            backtrack(nums, n, ind+1, res);
            Collections.swap(nums, ind, i);
        }
    }
}

File: LeetCode/Medium/Permutations/C++/permutations.cc
typedef vector<int> vi;
typedef vector<vi> vii;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vii res;
        backtrack(nums, res, nums.size(), 0);
        return res;
    }

    void backtrack(vi& nums, vii& res, int n, int ind) {
        if (ind == n) {
            res.push_back(nums);
            return;
        }

        for (int i=ind; i<n; i++) {
            swap(nums[i], nums[ind]);
            backtrack(nums, res, n, ind+1);
            swap(nums[i], nums[ind]);
        }
    }
};
